Add l1norm and l2norm distances for vectors

mayya-sharipova opened the pull request:
Add L1norm - Manhattan distance
Add L2norm - Euclidean distance
relates to #37947

mayya-sharipova committed:
Add l1norm and l2norm distances for vectors

Add L1norm - Manhattan distance
Add L2norm - Euclidean distance
relates to #37947

elasticmachine commented:
Pinging @elastic/es-search

mayya-sharipova commented:
@cbuescher  Can you please review this PR when you have time. I have tried to address here all your comments from #40255

cbuescher reviewed (changes requested):
Thanks @mayya-sharipova, looks very good already. I left a few minor questions and some asks around testing, nothing that seems to hold up this PR for a long time though.

mayya-sharipova commented on docs/reference/query-dsl/script-score-query.asciidoc:
```diff
@@ -31,8 +30,7 @@ GET /_search
 --------------------------------------------------
-// CONSOLE
-// TEST[setup:twitter]
+// NOTCONSOLE
```
@cbuescher this doesn't work anymore as there could only one TEST SETUP per a doc file.  I have decided to use TEST SETUP for testing vectors, and unable to use setup with twitter data for this example, so I made it `NOTCONSOLE`

mayya-sharipova replied:
an alternative could be to create a separate doc file for vector functions. I can do this

cbuescher replied:
No, thats fine then. I see enough snippets that are tested later. One without testing if okay I think if the way to solve this involves too much work.

mayya-sharipova commented on docs/reference/query-dsl/script-score-query.asciidoc:
```diff
@@ -97,6 +95,40 @@ cosine similarity between a given query vector and document vectors.
+PUT my_index/_doc/1
+{
+  "my_dense_vector": [0.5, 10, 6],
+  "my_sparse_vector": {"2": 1.5, "15" : 2, "50": -1.1, "4545": 1.1}
```
No, they can only be integers. We [throw an exception](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/SparseVectorFieldMapper.java#L163) if a vector key is not convertible to an integer.

cbuescher replied:
Thanks, good to know.

mayya-sharipova commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/ScoreScriptUtils.java:
```diff
@@ -141,32 +288,16 @@ public double dotProductSparse(VectorScriptDocValues.SparseVectorScriptDocValues
-            //break vector into two arrays dims and values
-            int n = queryVector.size();
-            queryValues = new double[n];
-            queryDims = new int[n];
```
@cbuescher One of [your comments from the previous PR](https://github.com/elastic/elasticsearch/pull/40255#discussion_r267992293) was for *Sparse classes to share a common  abstract superclass for e.g. sharing the common code in the constructor.  I thought this is a very good point and redesigned all *Sparse classes to extend from `VectorSparseFunctions` to share a common code in the constructor.

cbuescher replied:
Great, got it, thanks. I just wasn't sure where this moved to, might have been in a differen PR or I missed it here..

mayya-sharipova committed:
Address Christoph's feedback

- organize vector functions as a separate doc
- increase precision in tests calculations
- add a separate test when sparse doc dims
are bigger and less than query vector dims

jtibshirani commented on docs/reference/vectors/vector-functions.asciidoc:
```diff
@@ -0,0 +1,283 @@
+--------------------------------------------------
+// CONSOLE
+
+NOTE: Unlike `cosineSimilarity` that represent similarity, `l1norm` and
```
I had a couple suggestions around this transformation:
- We could include a 1 in the denominator (`1/ l1norm(params.queryVector, doc['my_dense_vector'])`) to avoid division by 0 when a document vector matches the query exactly.
- Would it make sense to use this transformation in the search themselves, instead of just including it in a note? We could give a description of it in a callout, as we do for `cosineSimilarity` and `dotProduct`. It would be harder to miss that way, and I also think it makes the examples more helpful + realistic, since it seems very rare to score by distance descending.

mayya-sharipova replied:
@jtibshirani Thanks these are very good suggestions, I have addressed them in the last commit

jtibshirani replied:
Thanks @mayya-sharipova, the new commit you added looks good to me!

cbuescher commented on x-pack/plugin/vectors/src/test/java/org/elasticsearch/xpack/vectors/query/ScoreScriptUtilsTests.java:
```diff
@@ -128,21 +128,58 @@ public void testSparseVectorFunctionsSpecialCases() {
         L1NormSparse l1Norm = new L1NormSparse(queryVector);
         double result3 = l1Norm.l1normSparse(dvs);
-        assertEquals("l1normSparse result is not equal to the expected value!", 517.18, result3, 0.1);
+        assertEquals("l1normSparse result is not equal to the expected value!", 517.184, result3, 0.001);
```
I would have expeced the l1 norm to change by 11.5 when adding the "4546" doc vector dimentions of 11.5f? 
Could you re-check the expected value? I think the while-loop in the function might miss the last "biggest" value?
Same for the l2-norm also btw.

cbuescher replied:
Sorry, my bad. I now saw you took out the "11" and moved it to "4546" with the same value. All good.

cbuescher reviewed (approved):
LGTM

mayya-sharipova committed:
Made examples more realistic

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into l1l2-vector-distances

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/2
